internal/nearcache/record_store.go

```diff
@@ -17,45 +17,80 @@
 package nearcache
 
 import (
+	"fmt"
 	"math/bits"
 	"sync"
 	"sync/atomic"
 	"time"
 
+	ihzerrors "github.com/hazelcast/hazelcast-go-client/internal/hzerrors"
 	"github.com/hazelcast/hazelcast-go-client/internal/serialization"
 	"github.com/hazelcast/hazelcast-go-client/nearcache"
 )
 
+const (
+	RecordNotReserved     int64 = -1
+	RecordReadPermitted   int64 = -2
+	RecordStoreTimeNotSet int64 = -1
+	// see: com.hazelcast.internal.eviction.impl.strategy.sampling.SamplingEvictionStrategy#SAMPLE_COUNT
+	RecordStoreSampleCount = 15
+)
+
+// serialization.Data is not hashable, so using this to store keys.
+type dataString string
+
 type RecordStore struct {
-	stats            Stats
-	maxIdleMillis    int64
-	reservationID    int64
-	timeToLiveMillis int64
-	recordsMu        *sync.Mutex
-	records          map[interface{}]*Record
-	ss               *serialization.Service
-	valueConverter   nearCacheRecordValueConverter
-	estimator        nearCacheStorageEstimator
+	stats             nearcache.Stats
+	maxIdleMillis     int64
+	reservationID     int64
+	timeToLiveMillis  int64
+	recordsMu         *sync.RWMutex
+	records           map[interface{}]*Record
+	ss                *serialization.Service
+	valueConverter    nearCacheRecordValueConverter
+	estimator         nearCacheStorageEstimator
+	staleReadDetector *StaleReadDetector
+	evictionDisabled  bool
+	maxSize           int
+	cmp               nearcache.EvictionPolicyComparator
 }
 
 func NewRecordStore(cfg *nearcache.Config, ss *serialization.Service, rc nearCacheRecordValueConverter, se nearCacheStorageEstimator) RecordStore {
-	stats := Stats{
+	stats := nearcache.Stats{
 		CreationTime: time.Now(),
 	}
 	return RecordStore{
-		recordsMu:        &sync.Mutex{},
+		recordsMu:        &sync.RWMutex{},
 		records:          map[interface{}]*Record{},
 		maxIdleMillis:    int64(cfg.MaxIdleSeconds * 1000),
 		ss:               ss,
 		timeToLiveMillis: int64(cfg.TimeToLiveSeconds * 1000),
 		valueConverter:   rc,
 		estimator:        se,
 		stats:            stats,
+		evictionDisabled: cfg.Eviction.EvictionPolicy() == nearcache.EvictionPolicyNone,
+		maxSize:          cfg.Eviction.Size(),
+		cmp:              getEvictionPolicyComparator(&cfg.Eviction),
 	}
 }
 
+func (rs *RecordStore) Clear() {
+	// port of: com.hazelcast.internal.nearcache.impl.store.AbstractNearCacheRecordStore#clear
+	// checkAvailable() does not apply since rs.records is always created
+	rs.recordsMu.Lock()
+	size := len(rs.records)
+	rs.records = nil
+	rs.recordsMu.Unlock()
+	atomic.StoreInt64(&rs.stats.OwnedEntryCount, 0)
+	atomic.StoreInt64(&rs.stats.OwnedEntryMemoryCost, 0)
+	atomic.AddInt64(&rs.stats.Invalidations, int64(size))
+	rs.incrementInvalidationRequests()
+}
+
 func (rs *RecordStore) Get(key interface{}) (value interface{}, found bool, err error) {
 	// checkAvailable() does not apply since rs.records is always created
+	rs.recordsMu.RLock()
+	defer rs.recordsMu.RUnlock()
```

Comment:
I think we can shrink the locked range by moving these lines but I don't know if performance will be noticably better